File: pipeline/taxonomy.py
"""Canonical editorial taxonomy for Web intelligence."""
from __future__ import annotations

from enum import StrEnum


class Track(StrEnum):
    POLICY_REGULATION = "policy-regulation"
    LITIGATION_LEGAL = "litigation-legal"
    SEARCH_DISCOVERY = "search-discovery"
    CRAWLER_CONTROLS = "crawler-controls"
    AGENTIC_WEB = "agentic-web"
    LICENSING_MONETIZATION = "licensing-monetization"
    STANDARDS_PROTOCOLS = "standards-protocols"
    ASSET_RIGHTS = "asset-rights"
    MEASUREMENT_ECONOMICS = "measurement-economics"


class Front(StrEnum):
    ACCESS_DISCOVERY = "access-discovery"
    AGENTS_IDENTITY = "agents-identity"
    RIGHTS_MARKETS = "rights-markets"
    GOVERNANCE_LAW = "governance-law"
    MEASUREMENT_ECONOMICS = "measurement-economics"


class SourceTier(StrEnum):
    PRIMARY = "primary"
    MEASUREMENT = "measurement"
    SPECIALIST = "specialist"
    COMMENTARY = "commentary"


class SourceRole(StrEnum):
    PLATFORM_DOCS = "platform-docs"
    STANDARDS = "standards"
    REGULATOR = "regulator"
    LEGAL_PRIMARY = "legal-primary"
    PUBLISHER = "publisher"
    INFRASTRUCTURE = "infrastructure"
    MEASUREMENT = "measurement"
    REPORTING = "reporting"


FRONT_TRACKS: dict[Front, tuple[Track, ...]] = {
    Front.ACCESS_DISCOVERY: (
        Track.CRAWLER_CONTROLS,
        Track.SEARCH_DISCOVERY,
    ),
    Front.AGENTS_IDENTITY: (
        Track.AGENTIC_WEB,
        Track.STANDARDS_PROTOCOLS,
    ),
    Front.RIGHTS_MARKETS: (
        Track.LICENSING_MONETIZATION,
        Track.ASSET_RIGHTS,
    ),
    Front.GOVERNANCE_LAW: (
        Track.POLICY_REGULATION,
        Track.LITIGATION_LEGAL,
    ),
    Front.MEASUREMENT_ECONOMICS: (Track.MEASUREMENT_ECONOMICS,),
}


def validate_tracks(primary_track: Track, tracks: list[Track]) -> None:
    """Validate an event's primary and secondary track assignments."""
    if not tracks:
        raise ValueError("tracks must not be empty")
    if len(tracks) != len(set(tracks)):
        raise ValueError("tracks must not contain duplicates")
    if primary_track not in tracks:
        raise ValueError("primary_track must appear in tracks")

File: site/src/lib/taxonomy.ts
export const TRACKS = [
  "policy-regulation",
  "litigation-legal",
  "search-discovery",
  "crawler-controls",
  "agentic-web",
  "licensing-monetization",
  "standards-protocols",
  "asset-rights",
  "measurement-economics",
] as const;

export type Track = (typeof TRACKS)[number];

export const FRONTS = [
  "access-discovery",
  "agents-identity",
  "rights-markets",
  "governance-law",
  "measurement-economics",
] as const;

export type Front = (typeof FRONTS)[number];

export const FRONT_TRACKS: Record<Front, readonly Track[]> = {
  "access-discovery": ["crawler-controls", "search-discovery"],
  "agents-identity": ["agentic-web", "standards-protocols"],
  "rights-markets": ["licensing-monetization", "asset-rights"],
  "governance-law": ["policy-regulation", "litigation-legal"],
  "measurement-economics": ["measurement-economics"],
};

export const TRACK_LABELS: Record<Track, string> = {
  "policy-regulation": "Policy & Regulation",
  "litigation-legal": "Litigation & Legal",
  "search-discovery": "Search & AI Discovery",
  "crawler-controls": "Crawlers & Content Controls",
  "agentic-web": "Agentic Web",
  "licensing-monetization": "Licensing & Monetization",
  "standards-protocols": "Standards & Protocols",
  "asset-rights": "Asset Rights",
  "measurement-economics": "Measurement & Economics",
};

export const FRONT_LABELS: Record<Front, string> = {
  "access-discovery": "Access & Discovery",
  "agents-identity": "Agents & Identity",
  "rights-markets": "Rights & Markets",
  "governance-law": "Governance & Law",
  "measurement-economics": "Measurement & Economics",
};

const flattened = FRONTS.flatMap((front) => FRONT_TRACKS[front]);
if (flattened.length !== TRACKS.length || new Set(flattened).size !== TRACKS.length) {
  throw new Error("Every track must belong to exactly one public front");
}

export function frontForTrack(track: Track): Front {
  const front = FRONTS.find((candidate) => FRONT_TRACKS[candidate].includes(track));
  if (!front) throw new Error(`No front configured for track: ${track}`);
  return front;
}

export type LegacyPillar = "crawler" | "ecosystem" | "agent";

export function legacyPillarForTracks(tracks: readonly Track[]): LegacyPillar {
  if (tracks.includes("crawler-controls")) return "crawler";
  if (tracks.includes("agentic-web") || tracks.includes("standards-protocols")) return "agent";
  return "ecosystem";
}
